core/src/main/java/org/apache/calcite/sql/SqlCall.java

```diff
@@ -140,7 +140,7 @@ public int operandCount() {
     final SqlDialect dialect = writer.getDialect();
     if (leftPrec > operator.getLeftPrec()
         || (operator.getRightPrec() <= rightPrec && (rightPrec != 0))
-        || writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION)) {
+        || (writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION))) {
```

Comment:
There is an open TODO to bug the calcite people about this, but based on my understanding, this is the correct grouping

Reply:
So, essentially, && has higher precedence then ||, so the addition of parenthesis here:
`(writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION)`

Doesn't change anything. However, based on my understanding of the overall logic, I think the correct grouping should be around the first three expressions as so:

`(leftPrec > operator.getLeftPrec()
        || (operator.getRightPrec() <= rightPrec && (rightPrec != 0))
        || (writer.isAlwaysUseParentheses()) && isA(SqlKind.EXPRESSION)`

Because it doesn't make sense to put parenthesis around something that isn't an expression. However, I wanted to bug the Calcite people first, just in case there is something that I'm not seeing.